### PR TITLE
enhancement(clickhouse sink): Add `ArrowStream` format

### PR DESCRIPTION
**Pull request by benjamin-awd**
## Summary
This PR adds support for the `ArrowStream` format to Clickhouse.

## Vector configuration

```yaml
 clickhouse:
    type: clickhouse
    inputs: [normalize]
    database: mydatabase
    table: logs
    endpoint: ${CLICKHOUSE_HOST}
    format: arrow_stream
    batch_encoding:
      codec: arrow_stream
      allow_nullable_fields: true
```

## How did you test this PR?

Ran against local Clickhouse instance, integration tests

## Change Type
- [ ] Bug fix
- [x] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?
- [ ] Yes
- [x] No

## Does this PR include user facing changes?

- [x] Yes. Please add a changelog fragment based on our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [ ] No. A maintainer will apply the `no-changelog` label to this PR.

## References

- Closes https://github.com/vectordotdev/vector/issues/24074

## Notes
- Please read our [Vector contributor resources](https://github.com/vectordotdev/vector/tree/master/docs#getting-started).
- Do not hesitate to use `@vectordotdev/vector` to reach out to us regarding this PR.
- Some CI checks run only after we manually approve them.
  - We recommend adding a `pre-push` hook, please see [this template](https://github.com/vectordotdev/vector/blob/master/CONTRIBUTING.md#Pre-push).
  - Alternatively, we recommend running the following locally before pushing to the remote branch:
    - `make fmt`
    - `make check-clippy` (if there are failures it's possible some of them can be fixed with `make clippy-fix`)
    - `make test`
- After a review is requested, please avoid force pushes to help us review incrementally.
  - Feel free to push as many commits as you want. They will be squashed into one before merging.
  - For example, you can run `git merge origin master` and `git push`.
- If this PR introduces changes Vector dependencies (modifies `Cargo.lock`), please
  run `make build-licenses` to regenerate the [license inventory](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv) and commit the changes (if any). More details [here](https://crates.io/crates/dd-rust-license-tool).

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -406,7 +407,7 @@ redis = { version = "0.32.4", default-features = false, features = ["connection-
 regex.workspace = true
 roaring = { version = "0.11.2", default-features = false, features = ["std"], optional = true }
 rumqttc = { version = "0.24.0", default-features = false, features = ["use-rustls"], optional = true }
-rust_decimal = { workspace = true, optional = true }
+rust_decimal = { version = "1.37.0", default-features = false, features = ["std"], optional = true }
```
Comment (thomasqueirozb) [inline Cargo.toml]: Looks like `rust_decimal` is already using this version in `lib/codecs/Cargo.toml`. We should update the workspace reference to use `1.37.0` and keep using workspace here. Additionally `lib/codecs/Cargo.toml` should also use the `workspace` version of `rust_decimal`

File: src/sinks/clickhouse/config.rs [hunk 2]
```diff
@@ -249,6 +260,127 @@ impl SinkConfig for ClickhouseConfig {
...
+            if self.format != Format::ArrowStream {
+                return Err(format!(
+                    "'batch_encoding' is only compatible with 'format: arrow_stream'. \
+                         Found 'format: {}'.",
+                    match self.format {
+                        Format::JsonEachRow => "json_each_row",
+                        Format::JsonAsObject => "json_as_object",
+                        Format::JsonAsString => "json_as_string",
+                        Format::ArrowStream => "arrow_stream",
+                    }
```
Comment (thomasqueirozb) [inline src/sinks/clickhouse/config.rs]: Isn't this just the serde representation of `Format`? Wondering if we should keep this as is or use `serde` here somehow to avoid drift

File: src/sinks/clickhouse/arrow_schema.rs [hunk 3]
```diff
@@ -0,0 +1,767 @@
+//! Schema fetching and Arrow type mapping for ClickHouse tables.
+
+use std::sync::Arc;
+
+use ::arrow::datatypes::{DataType, Field, Schema, TimeUnit};
```
Comment (thomasqueirozb) [inline src/sinks/clickhouse/arrow_schema.rs]:
```suggestion
use arrow::datatypes::{DataType, Field, Schema, TimeUnit};
```

Don't think the `::` prefix is needed

File: src/sinks/clickhouse/arrow_schema.rs [hunk 4]
```diff
@@ -0,0 +1,767 @@
...
+    table: &str,
+    auth: Option<&Auth>,
+) -> crate::Result<Arc<Schema>> {
+    // Query to get table schema
+    let query = format!(
+        "SELECT name, type \
+         FROM system.columns WHERE database = '{}' AND table = '{}' \
+         ORDER BY position FORMAT JSONEachRow",
+        database, table
+    );
```
Comment (thomasqueirozb) [inline src/sinks/clickhouse/arrow_schema.rs]: For an abundance of safety here I think we should have some kind of SQLi prevention. Something as simple as limiting database and table to ascii chars and `-, _` should be enough. I wouldn't be opposed to using some other crate here instead.

Note: I spent some time looking into if the `sqlx` crate is viable for this but it doesn't look like it is unfortunately. All query operations depend on a database connection.

File: src/sinks/clickhouse/arrow_schema.rs [hunk 5]
```diff
@@ -0,0 +1,767 @@
...
+        fields.push(Field::new(&column.name, arrow_type, nullable));
+    }
+
+    Ok(Arc::new(Schema::new(fields)))
+}
+
+/// Unwraps ClickHouse type modifiers like Nullable() and LowCardinality().
+/// Returns a tuple of (base_type, is_nullable).
+/// For example: "Nullable(LowCardinality(String))" -> ("String", true)
+fn unwrap_type_modifiers(ch_type: &str) -> (&str, bool) {
```
Comment (thomasqueirozb) [inline src/sinks/clickhouse/arrow_schema.rs]: Is it possible to have `LowCardinality(Nullable(String))`? If so this logic fails

File: src/sinks/clickhouse/arrow_schema.rs [hunk 6]
```diff
@@ -0,0 +1,767 @@
...
+    let mut base = ch_type;
+    let mut is_nullable = false;
+
+    // Check for Nullable wrapper
+    if let Some(inner) = base.strip_prefix("Nullable(") {
+        is_nullable = true;
+        base = inner.strip_suffix(')').unwrap_or(inner);
```
Comment (thomasqueirozb) [inline src/sinks/clickhouse/arrow_schema.rs]: The `.unwrap_or`s in this function assume that incoming data is always closed like `Nullable(LowCardinality(String))`, but it theoretically would work with malformed data like `Nullable(LowCardinality(String` or `Nullable(LowCardinality(String)`. Realistically, do we want to fail on these or should we just silently accept? I would think that this is not a very likely scenario either way since, if I'm understanding this correctly, these type modifiers come directly from Clickhouse

File: src/sinks/clickhouse/arrow/parser.rs [hunk 7]
```diff
@@ -0,0 +1,767 @@
...
+        // Complex types
+        _ if base_type.starts_with("Array") => {
+            return Err(unsupported(ch_type, "Array"));
+        }
+        _ if base_type.starts_with("Tuple") => {
+            return Err(unsupported(ch_type, "Tuple"));
+        }
+        _ if base_type.starts_with("Map") => {
+            return Err(unsupported(ch_type, "Map"));
+        }
```
Comment (thomasqueirozb) [inline src/sinks/clickhouse/arrow/parser.rs]: Out of scope: is this something you plan to add support for in the future? Is this something that would add much value?

File: src/sinks/clickhouse/arrow/parser.rs [hunk 8]
```diff
@@ -0,0 +1,767 @@
...
+        return (&input[..i], &input[i..]);
+    }
+    (input, "")
+}
+
+/// Parses comma-separated arguments from a parenthesized string.
+/// Input: "(arg1, arg2, arg3)" -> Output: Ok(vec!["arg1".to_string(), "arg2".to_string(), "arg3".to_string()])
+/// Returns an error if parentheses are malformed.
+fn parse_args(input: &str) -> Result<Vec<String>, String> {
```
Comment (thomasqueirozb) [inline src/sinks/clickhouse/arrow/parser.rs:151]: Not saying there is anything wrong with this implementation, but usually manually keeping track of nested quotes and paren depth is not trivial and very error prone. Ideally we'd have robust tests for this. Another idea is to parse this using `nom` also, this way you can get rid of the state machine but it could also be that the nom code wouldn't be much better.

<details>
<summary>LLM generated nom parser - not tested at all</summary>

```rs
/// Parses comma-separated arguments from a parenthesized string.
/// Input: "(arg1, arg2, arg3)" -> Output: Ok(vec!["arg1".to_string(), "arg2".to_string(), "arg3".to_string()])
/// Returns an error if parentheses are malformed.
fn parse_args(input: &str) -> Result<Vec<String>, String> {
    use nom::{
        IResult,
        branch::alt,
        bytes::complete::{tag, take_until},
        character::complete::{char, multispace0},
        combinator::{map, recognize},
        multi::separated_list0,
        sequence::{delimited, pair, preceded},
    };

    fn quoted_string(input: &str) -> IResult<&str, &str> {
        delimited(char('\''), take_until("'"), char('\''))(input)
    }

    fn nested_parens(input: &str) -> IResult<&str, &str> {
        // This handles nested structures like "Array(Int32)"
        recognize(pair(
            take_until("("),
            delimited(char('('), nested_arg, char(')')),
        ))(input)
    }

    fn nested_arg(input: &str) -> IResult<&str, &str> {
        recognize(alt((
            quoted_string,
            nested_parens,
            nom::bytes::complete::is_not(",)"),
        )))(input)
    }

    fn single_arg(input: &str) -> IResult<&str, &str> {
        preceded(
            multispace0,
            alt((
                quoted_string,
                nested_parens,
                nom::bytes::complete::is_not(",)"),
            )),
        )(input)
    }

    fn args_parser(input: &str) -> IResult<&str, Vec<&str>> {
        delimited(
            char('('),
            separated_list0(
                delimited(multispace0, char(','), multispace0),
                map(single_arg, str::trim),
            ),
            preceded(multispace0, char(')')),
        )(input)
    }

    let trimmed = input.trim();
    match args_parser(trimmed) {
        Ok(("", args)) => Ok(args.into_iter().map(String::from).collect()),
        Ok((remaining, _)) => Err(format!("Unexpected trailing input: '{}'", remaining)),
        Err(e) => Err(format!("Failed to parse arguments: {}", e)),
    }
}
```

</details>

I also think this is would break if something like this came back: `('arg\'\)')`. Since this is data coming straight from clickhouse I'm not sure if this is a problem but if any user-controlled data is involved this could potentially happen.


Everything I said in here basically boils down to:
1. Function could maybe be improved using `nom`.
2. Needs tests.
3. User controlled data can result in an `Err`. (Not sure if applicable, we can also just `Err` if something weird comes in)

File: src/sinks/clickhouse/arrow/parser.rs [hunk 9]
```diff
@@ -0,0 +1,767 @@
...
+        .map(|(precision, scale)| {
+            if precision <= DECIMAL128_PRECISION {
+                DataType::Decimal128(precision, scale)
+            } else {
+                DataType::Decimal256(precision, scale)
+            }
+        })
+        .ok_or_else(|| {
+            format!(
+                "Could not parse Decimal type '{}'. Valid formats: Decimal(P,S), Decimal32(S), Decimal64(S), Decimal128(S), Decimal256(S)",
```
Comment (thomasqueirozb) [inline src/sinks/clickhouse/arrow/parser.rs]:
```suggestion
                "Could not parse Decimal type '{}'. Valid formats: Decimal(P, S), Decimal32(S), Decimal64(S), Decimal128(S), Decimal256(S)",
```

Either this or the docstring needs to be changed

File: src/sinks/clickhouse/arrow/parser.rs [hunk 10]
```diff
@@ -0,0 +1,767 @@
...
+        "Float32" => DataType::Float32,
+        "Float64" => DataType::Float64,
+
+        // Boolean
+        "Bool" => DataType::Boolean,
+
+        // Date and time types (timezones not currently handled, defaults to UTC)
+        "Date" | "Date32" => DataType::Date32,
+        "DateTime" => DataType::Timestamp(TimeUnit::Second, None),
+        _ if base_type.starts_with("DateTime64") => parse_datetime64_precision(base_type)?,
```
Comment (thomasqueirozb) [inline src/sinks/clickhouse/arrow/parser.rs]: Using `starts_with` here is prone to someone using `DateTime64FooBarBaz` and getting back a DateTime64. Again I'm worried about two things here: first is untrusted input, but not sure if applicable; second is that if some other `DateTime64` type is added like `DateTime64UTC` or something like that then this becomes a real problem. You already implemented `extract_identifier`, so the simplest solution is to move all these to a `_ => {` block and matching on `extract_identifier().0` instead of using `str::starts_with`.

This is of course relevant in the other matches that are using `starts_with`

File: src/sinks/clickhouse/config.rs [hunk 11]
```diff
@@ -249,6 +260,127 @@ impl SinkConfig for ClickhouseConfig {
...
+            client.clone(),
+            endpoint,
+            database_str.to_string(),
+            table_str.to_string(),
+            auth.cloned(),
+        ));
+
+        let mut arrow_config = ArrowStreamSerializerConfig::with_provider(provider);
+        // Preserve allow_nullable_fields setting from the user's config
+        arrow_config.allow_nullable_fields = base_config.allow_nullable_fields;
```
Comment (thomasqueirozb) [inline src/sinks/clickhouse/config.rs]: Having the one allow_nullable_fields as an "exception" that we pass along to the newly created `arrow_config` can be a source of bugs in the future if more fields to `ArrowStreamSerializerConfig`. I think we would be better off modifying the original `base_config` by removing schema and inserting a schema provider instead, either directly by doing `base_config.schema = None` and `base_config.schema_provider = Some(provider)` or by doing something like `base_config.with_schema_provider(provider)`.

File: lib/codecs/src/encoding/format/arrow.rs [hunk 12]
```diff
@@ -45,6 +55,10 @@ pub struct ArrowStreamSerializerConfig {
     #[serde(default)]
     #[configurable(metadata(docs::examples = true))]
     pub allow_nullable_fields: bool,
+
+    /// Schema provider for lazy schema loading.
+    #[serde(skip)]
+    schema_provider: Option<Arc<dyn SchemaProvider>>,
```
Comment (thomasqueirozb) [inline lib/codecs/src/encoding/format/arrow.rs]: Is Arc necessary? I'd think this would work with `Box<>` too

File: lib/codecs/src/encoding/format/arrow.rs [hunk 13]
```diff
@@ -45,6 +55,10 @@ pub struct ArrowStreamSerializerConfig {
     #[serde(default)]
     #[configurable(metadata(docs::examples = true))]
     pub allow_nullable_fields: bool,
+
+    /// Schema provider for lazy schema loading.
```
Comment (pront) [inline lib/codecs/src/encoding/format/arrow.rs]: Are we anticipating runtime schema changes? If not, I would move schema fetching entirely into config build phase.

File: Cargo.toml (hunk 1 above)
Comment (benjamin-awd) [reply Cargo.toml]: Changes made in e7b89c0bd7aa8fcfb04b217f3c289e584f9e50d2

File: src/sinks/clickhouse/config.rs (hunk 2 above)
Comment (benjamin-awd) [reply src/sinks/clickhouse/config.rs]: Decided to simplify the validation a bit, let me know if it makes sense 9e9392669

If there are additional batch formats in the future, it might make sense to create a separate enum but I think this should be good enough for now

File: src/sinks/clickhouse/arrow_schema.rs (hunk 3 above)
Comment (benjamin-awd) [reply src/sinks/clickhouse/arrow_schema.rs]: 85d7aa33f

File: src/sinks/clickhouse/arrow_schema.rs (hunk 4 above)
Comment (benjamin-awd) [reply src/sinks/clickhouse/arrow_schema.rs]: I think the ideal solution would be to use the official Clickhouse rust client:
```rust
use clickhouse::sql::Identifier;

client.query("SELECT * FROM ?").bind(Identifier("table_name"));
```

but I think it only makes sense as part of a long term refactor of the entire sink to move away from HTTP. This would also be nice since it allows query level settings to be passed directly via the client instead of defining them in the config
```rust
    let client = Client::default()
        .with_url("http://localhost:8123")
        .with_option("async_insert", "1")
        .with_option("wait_for_async_insert", "0");
```

Comment (benjamin-awd) [reply src/sinks/clickhouse/arrow_schema.rs]: Added more robust handling: 4e101ac00

File: src/sinks/clickhouse/arrow_schema.rs (hunk 5 above)
Comment (benjamin-awd) [reply src/sinks/clickhouse/arrow_schema.rs]: Ah it should actually be `LowCardinality(Nullable(String))`

The inverse is not possible:

```
CREATE TABLE lc_test
(
    `id` UInt16,
    `col` Nullable(LowCardinality(String))
)
ENGINE = MergeTree
ORDER BY id

Received exception from server (version 25.4.5):
Code: 43. DB::Exception: Received from localhost:9000. DB::Exception: Nested type LowCardinality(String) cannot be inside Nullable type. (ILLEGAL_TYPE_OF_ARGUMENT)
```

File: src/sinks/clickhouse/arrow/parser.rs (hunk 7 above)
Comment (benjamin-awd) [reply src/sinks/clickhouse/arrow/parser.rs]: Yes, eventually -- but it's a little complex and will involve changes to the Arrow encoder logic as well

Decided to elide it from this PR to make it a bit easier to review

File: src/sinks/clickhouse/arrow_schema.rs (hunk 5 above)
Comment (benjamin-awd) [reply src/sinks/clickhouse/arrow_schema.rs]: Decided to refactor the type handling: 4211ce349 (also fixed the ordering issue)

File: src/sinks/clickhouse/arrow_schema.rs (hunk 6 above)
Comment (benjamin-awd) [reply src/sinks/clickhouse/arrow_schema.rs]: Should be addressed by 4211ce349, but yes these types come directly from Clickhouse and are validated by the engine during the DDL stage

If a column is defined with bad syntax, it will fail

File: src/sinks/clickhouse/arrow/parser.rs (hunk 8 above)
Comment (benjamin-awd) [reply src/sinks/clickhouse/arrow/parser.rs:151]: The `nom` idea is pretty cool, but it's quite verbose and I think the current approach is still sustainable for now. Regardless, we still need a looooot of custom logic for this

> User controlled data can result in an Err. (Not sure if applicable, we can also just Err if something weird comes in)

We don't need to worry about user controlled data, since it should be validated by Clickhouse already. 

> Needs tests 

There are some tests already for parse_args and parse_ch_type, but there there are probably edge cases that might pop up -- I think it would be advisable to flag this feature as beta in the docs 

I was looking at [clickhouse-arrow](https://github.com/GeorgeLeePatterson/clickhouse-arrow/blob/1b4716615f9903b45efc402c91a6a3f4243a4f74/clickhouse-arrow/src/native/types/deserialize.rs#L525) for inspiration but it is also fairly complex

File: src/sinks/clickhouse/arrow/parser.rs (hunk 9 above)
Comment (benjamin-awd) [reply src/sinks/clickhouse/arrow/parser.rs]: Updated the error message: 1826aa3d1

File: src/sinks/clickhouse/arrow/parser.rs (hunk 10 above)
Comment (benjamin-awd) [reply src/sinks/clickhouse/arrow/parser.rs]: Implemented by 1db0f07d5

File: src/sinks/clickhouse/config.rs (hunk 11 above)
Comment (benjamin-awd) [reply src/sinks/clickhouse/config.rs]: Good catch, changes made: [e1b22a1d8](https://github.com/vectordotdev/vector/commit/e1b22a1d8)

File: lib/codecs/src/encoding/format/arrow.rs (hunk 12 above)
Comment (benjamin-awd) [reply lib/codecs/src/encoding/format/arrow.rs]: Since the schema provider is currently only being used once at startup before being discarded I think it's fine to go with `Box`

File: lib/codecs/src/encoding/format/arrow.rs (hunk 13 above)
Comment (benjamin-awd) [reply lib/codecs/src/encoding/format/arrow.rs]: Made changes in aa69c1c89  -- I think we only need this at runtime if the target database/table is dynamic and determined by a VRL remap transform

For now I think it makes sense to keep things simple and just keep schema fetching in the config as you mentioned  👍

File: lib/codecs/src/encoding/format/arrow.rs (hunk 12 above)
Comment (benjamin-awd) [reply lib/codecs/src/encoding/format/arrow.rs]: Should be handled by 834ce89b9..aa69c1c89

**Comment by akutta** [comment]
```
2025-12-15T13:37:08.641679Z ERROR vector::topology::builder: Configuration error. error=Sink "clickhouse": Failed to fetch schema for logs.logs: Failed to fetch schema from provider: Failed to convert column 'other_attributes': Unknown ClickHouse type 'JSON'. This type cannot be automatically converted.. internal_log_rate_limit=false
```

Per the docs; this is a limitation of the Arrow format:

```
Unsupported Arrow data types:

FIXED_SIZE_BINARY
JSON
UUID
ENUM.
```

So it's not fully backwards compatible with the JSON encoder.  Would make sense to call this out in the documentation.

File: src/sinks/clickhouse/arrow/parser.rs (hunk 10 above)
Comment (thomasqueirozb) [reply src/sinks/clickhouse/arrow/parser.rs]: Very nice

File: Cargo.toml (hunk 1 above)
Comment (thomasqueirozb) [reply Cargo.toml]:
```suggestion
rust_decimal = { workspace = true, optional = true }
```

Still need to revert changes to this line

File: src/sinks/clickhouse/arrow/parser.rs (hunk 8 above)
Comment (thomasqueirozb) [reply src/sinks/clickhouse/arrow/parser.rs:151]: > but it's quite verbose

yep I agree with that.

Since we don't have to worry about invalid user controlled data then we can ignore those weird strings I mentioned and just return `Err`

File: src/sinks/clickhouse/arrow_schema.rs (hunk 4 above)
Comment (thomasqueirozb) [reply src/sinks/clickhouse/arrow_schema.rs]: Didn't know that `&param_XYZ` was supported, newer param construction is very nice.

> I think the ideal solution would be to use the official Clickhouse rust client

Moving over to a rust based clickhouse client could prove to be a good alternative in the future, but I think some discussion is needed. These big updates can cause regressions or have unintended functionality

File: src/sinks/clickhouse/arrow_schema.rs (hunk 5 above)
Comment (thomasqueirozb) [reply src/sinks/clickhouse/arrow_schema.rs]: Refactor looks good. I don't think we need to worry about some weird data like Nullable(Nullable()) coming in since it's coming directly from Clickhouse. If it does then oh well and we'll hit an error somewhere

File: Cargo.toml (hunk 1 above)
Comment (thomasqueirozb) [reply Cargo.toml]: See d60aac60425528ded19b7bce6cfa5ce315fd6a07

File: src/sinks/clickhouse/arrow/parser.rs (hunk 8 above)
Comment (thomasqueirozb) [reply src/sinks/clickhouse/arrow/parser.rs:151]: Quick note: since we can assume that the format is coming from Clickhouse and that it is valid already this implementation is just fine. Nom would be able to provide error handling and a lot of niceties, all of which we don't need here atm

File: Cargo.toml (hunk 1 above)
Comment (benjamin-awd) [reply Cargo.toml]: Thanks @thomasqueirozb! Still trying to wrap my head around Rust dependencies

File: Cargo.toml [hunk 14]
```diff
@@ -851,7 +852,7 @@ sinks-azure_blob = ["dep:azure_core", "dep:azure_identity", "dep:azure_storage",
 sinks-azure_monitor_logs = []
 sinks-blackhole = []
 sinks-chronicle = []
-sinks-clickhouse = []
+sinks-clickhouse = ["dep:arrow", "dep:arrow-schema", "dep:rust_decimal", "codecs-arrow"]
```
Comment (thomasqueirozb) [inline Cargo.toml]: some of these are duplicated from codecs-arrow and could be removed

File: src/sinks/clickhouse/arrow/schema.rs [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+            database,
+            table,
+            auth,
+        }
+    }
+}
+
+#[async_trait]
+impl SchemaProvider for ClickHouseSchemaProvider {
+    async fn get_schema(&self) -> Result<Arc<Schema>, ArrowEncodingError> {
```
Comment (thomasqueirozb) [inline src/sinks/clickhouse/arrow/schema.rs]: Not necessarily wrong, but I don't see an use case for `Arc`. Looking at everything it looks like `ArrowStreamSerializerConfig` could take in the owned `Schema` value without any significant changes

File: src/sinks/clickhouse/integration_tests.rs [hunk 16]
```diff
@@ -468,3 +469,139 @@ struct Stats {
...
+    let host = clickhouse_address();
+
+    let mut batch = BatchConfig::default();
+    batch.max_events = Some(5);
+
+    let config = ClickhouseConfig {
+        endpoint: host.parse().unwrap(),
+        table: table.clone().try_into().unwrap(),
+        compression: Compression::None,
+        format: crate::sinks::clickhouse::config::Format::ArrowStream,
```
Comment (thomasqueirozb) [inline src/sinks/clickhouse/integration_tests.rs:487]: If I'm understanding this correctly, these tests would pass without ArrowStream format right? It'd be nice to add a test that passes using arrow and doesn't otherwise - not sure how feasible it is though and we shouldn't add to many hacks to get this to work

File: Cargo.toml (hunk 14 above)
Comment (benjamin-awd) [reply Cargo.toml]: 162181416

File: src/sinks/clickhouse/arrow/schema.rs (hunk 15 above)
Comment (benjamin-awd) [reply src/sinks/clickhouse/arrow/schema.rs]: I think it is indeed cleaner to store the Arc at the serializer level instead of at the config level -- changes made 4efff4856

NB: [RecordBatch](https://arrow.apache.org/rust/arrow/record_batch/struct.RecordBatch.html) requires `Arc<Schema>` so we still need to convert the schema somewhere

File: src/sinks/clickhouse/integration_tests.rs (hunk 16 above)
Comment (benjamin-awd) [reply src/sinks/clickhouse/integration_tests.rs:487]: The test will fail if the ArrowStream format is not explicitly passed via HTTP -- right now we're doing the equivalent of something like this (using JSONEachRow as an example):

```bash
# creating a dummy table
echo "CREATE TABLE test (foo String) ENGINE = Memory" \
  | curl 'http://localhost:8123/' --data-binary @-

# this will succeed
echo '{"foo":"bar"}' \
  | curl 'http://localhost:8123/?query=INSERT%20INTO%20test%20FORMAT%20JSONEachRow' \
         --data-binary @-
```

If we specify `ArrowStream` as a format but pass a JSON payload, Clickhouse will reject the INSERT:
```bash
echo '{"foo":"bar"}' \
  | curl 'http://localhost:8123/?query=INSERT%20INTO%20test%20FORMAT%20ArrowStream' \
         --data-binary @- 
Code: 1002. DB::Exception: Error while opening a table: Invalid: Expected to read 1868964475 metadata bytes, but only read 10: While executing ArrowBlockInputFormat. (UNKNOWN_EXCEPTION) (version 25.12.1.649 (official build))
```

I don't think there would be much value testing this, since we'd just be checking that Clickhouse is able to differentiate between different formats

The current tests make sure that:
1. We can serialize a batch of events into the Arrow IPC format (i.e. RecordBatch)
2. We can connect to Clickhouse and successfully fetch a schema
3. ClickHouse types are correctly mapped to Arrow types
4. Data from ClickHouse matches what we expect to see

There's probably more room for "unhappy" tests though but I think this should be okay for now

Comment (thomasqueirozb) [reply src/sinks/clickhouse/integration_tests.rs:487]: I didn't know that was the case. It's fine to leave as is then. Thanks for the explanation!
